x-pack/apm-server/sampling/eventstorage/storage_manager.go

```diff
@@ -341,30 +397,64 @@ func (sm *StorageManager) WriteSubscriberPosition(data []byte) error {
 // NewUnlimitedReadWriter returns a read writer with no storage limit.
 // For testing only.
 func (sm *StorageManager) NewUnlimitedReadWriter() StorageLimitReadWriter {
-	return sm.NewReadWriter(0)
+	return sm.NewReadWriter(0, 0)
 }
 
-// NewReadWriter returns a read writer with storage limit.
-func (sm *StorageManager) NewReadWriter(storageLimit uint64) StorageLimitReadWriter {
+// NewReadWriter returns a read writer configured with storage limit and disk threshold ratio.
+func (sm *StorageManager) NewReadWriter(storageLimit uint64, diskThresholdRatio float64) StorageLimitReadWriter {
 	splitRW := SplitReadWriter{
 		eventRW:    sm.eventStorage.NewReadWriter(),
 		decisionRW: sm.decisionStorage.NewReadWriter(),
 	}
 
-	dbStorageLimit := func() uint64 {
-		return storageLimit
-	}
-	if storageLimit == 0 {
-		sm.logger.Infof("setting database storage limit to unlimited")
+	// dbStorageLimit returns max size of db in bytes.
+	// If size of db exceeds dbStorageLimit, writes should be rejected.
+	var dbStorageLimit func() uint64
+	if sm.diskStatFailed.Load() && storageLimit == 0 && diskThresholdRatio > 0 {
+		dbStorageLimit = func() uint64 {
+			return dbStorageLimitFallback
+		}
+		sm.logger.Warnf("failed to get disk usage; overriding database storage limit to fallback default of %0.1fgb", float64(dbStorageLimitFallback)/gb)
 	} else {
-		sm.logger.Infof("setting database storage limit to %.1fgb", float64(storageLimit))
+		dbStorageLimit = func() uint64 {
+			return storageLimit
+		}
+		if storageLimit == 0 {
+			sm.logger.Infof("setting database storage limit to unlimited")
+		} else {
+			sm.logger.Infof("setting database storage limit to %0.1fgb", float64(storageLimit)/gb)
+		}
 	}
 
 	// To limit db size to storage_limit
 	dbStorageLimitChecker := NewStorageLimitCheckerFunc(sm.dbSize, dbStorageLimit)
 	dbStorageLimitRW := NewStorageLimitReadWriter("database storage limit", dbStorageLimitChecker, splitRW)
 
-	return dbStorageLimitRW
+	// diskThreshold returns max used disk space in bytes.
+	// If size of used disk space exceeds diskThreshold, writes should be rejected.
+	var diskThreshold func() uint64
+	if sm.diskStatFailed.Load() {
+		diskThreshold = func() uint64 {
+			return 0 // unlimited
+		}
+		sm.logger.Warnf("failed to get disk usage; disabling disk threshold check")
+	} else {
+		diskThreshold = func() uint64 {
+			return uint64(float64(sm.diskStat.total.Load()) * diskThresholdRatio)
+		}
+		// the total disk space could change in runtime, but it is still useful to print it out in logs.
+		sm.logger.Infof("setting disk threshold ratio to %.2f of total disk space of %0.1fgb", diskThresholdRatio, float64(sm.diskStat.total.Load())/gb)
+	}
+
+	// To limit actual disk usage percentage to diskThresholdRatio
+	diskThresholdChecker := NewStorageLimitCheckerFunc(sm.diskUsed, diskThreshold)
+	diskThresholdRW := NewStorageLimitReadWriter(
+		fmt.Sprintf("disk usage ratio exceeding threshold of %.2f", diskThresholdRatio),
```

Comment:
How is this used? The docs says that this is a `name` parameter

Reply:
It first started as a name of the StorageLimitReadWriter, but now I'm using it more like a description. It is prepended to the error. While changing it to "disk usage threshold" makes it become a name again, having the value of it seems to help understand, because `current: X, limit: Y` in the error message are absolute values and is not very easy to understand: `disk usage ratio exceeding threshold of 0.01: configured storage limit reached (current: 38973440, limit 21474836)`

Reply:
Hmm, tbh, it feels a bit of a smell. Maybe we can refactor it later to abstract the error outside `StorageLimitReadWriter` and remove the `name`.

Reply:
No need to do this now - we can do this in a future PR.

Reply:
updated it slightly to make the passed in argument more like a name. Abstracting the error outside StorageLimitReadWriter to create a DiskThresholdReadWriter is fine, but they would look too similar apart from the error / name.